api/src/main/java/jakarta/data/spi/provider/DataProvider.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.spi.provider;
+
+import java.lang.annotation.Annotation;
+import java.util.Set;
+
+import jakarta.data.exceptions.MappingException;
+
+/**
+ * <p>Jakarta Data providers can make themselves discoverable by
+ * Jakarta EE products via the Jakarta Data Provider SPI. This enables
+ * Jakarta EE products with CDI support to provide CDI dependency injection
+ * and other integration with CDI on behalf of Jakarta Data providers.
+ * Jakarta Data providers with their own custom dependency injection
+ * are not required to implement or provide the Jakarta Data Provider SPI.</p>
+ *
+ * <p>Jakarta Data providers that implement the Jakarta Data Provider SPI
+ * make an implementation of this interface available to the
+ * {@link java.util.ServiceLoader ServiceLoader} as follows.</p>
+ *
+ * <p>The {@code DataProvider} implementation and related classes are packaged
+ * within the Jakarta Data provider's JAR file, also including a file with the
+ * following name and location,</p>
+ *
+ * <pre>META-INF/services/jakarta.data.spi.provider.DataProvider</pre>
+ * 
+ * <p>The content of the above file must be one or more lines, each specifying
+ * the fully qualified name of a {@code DataProvider} implementation that is
+ * provided within the JAR file.</p>
+ *
+ * <p>Jakarta EE applications do not use the Jakarta Data Provider SPI.
+ * Jakarta EE products are not required to make this class available to
+ * applications.</p>
+ */
+public interface DataProvider {
+    /**
+     * <p>Provides an instance that implements the specified repository interface.</p>
+     *
+     * @param <R>                 interface class that defines the data repository.
+     * @param repositoryInterface the repository interface.
+     * @param entityClass         type of entity that the repository persists.
+     * @return repository instance. Never {@code null}.
+     * @throws MappingException for inconsistencies between the repository or
+     *         entity class and the database.
+     */
+    <R> R getRepository(Class<R> repositoryInterface, Class<?> entityClass) throws MappingException;
```

Comment:
```suggestion
<R> R getRepository(Class<R> repository) throws MappingException;
```

You don't need the second one, you can get it from the class:


```java
        Class<?> repository = PersonRepository.class;
        System.out.println("repository: " + repository);
        ParameterizedType param = (ParameterizedType) repository.getGenericInterfaces()[0];
        Type[] arguments = param.getActualTypeArguments();
        System.out.println("the " + param);
        System.out.println("the entity: " + arguments[0]);
        System.out.println("the id: " + arguments[1]);
```

Reply:
That's a very good point, the entityClass parameter is redundant because it can be obtained from the repository class,  so let's go with your change to remove it.